connector/src/main/scala/com/datastax/spark/connector/datasource/CassandraCatalog.scala

```diff
@@ -0,0 +1,431 @@
+    val colName = checkColumnName(name)
+    val primaryKeys = tableMeta.getPrimaryKey.asScala.map(_.getName.asInternal()).toSet
+    if (primaryKeys.contains(colName.asInternal())) {
+      throw new CassandraCatalogException(s"Cassandra cannot drop primary key columns: Tried to drop $colName")
+    }
+    colName
+  }
+
+  /**
+    * Limited to Adding and Removing Normal Columns and Setting Properties
+    *
+    * The API expects that if any change is rejected, we  should not apply
+    * any changes. This is basically impossible for us since we cannot batch
+    * together all of our DDL changes into a single CQL statement. To ameliorate
+    * this we perform as many checks as we can up front to try to avoid beginning
+    * our DDL requests which we know will fail.
+    *
+    * We break up DDL into 3 phases
+    * Set Properties
+    * Remove Columns
+    * Add Columns
+    */
+  override def alterTable(ident: Identifier, changes: TableChange*): Table = {
+    val protocolVersion = connector.withSessionDo(_.getContext.getProtocolVersion)
+    val tableMetadata = getTableMetaData(connector, ident)
+    val keyspace = tableMetadata.getKeyspace
+    val table = tableMetadata.getName
+
+    //Check for unsupported table changes
+    changes.foreach {
+      case add: AddColumn =>
+      case del: DeleteColumn =>
+      case setProp: TableChange.SetProperty =>
+      case other: TableChange => throw new CassandraCatalogException(s"Cassandra Catalog does not support Alter operation: ${other.getClass.getSimpleName}")
+    }
+
+    val propertiesToAdd = changes.collect { case setProperty: TableChange.SetProperty =>
+      setProperty
+    }
+    val columnsToRemove = changes.collect { case remove: DeleteColumn =>
+      checkRemoveNormalColumn(tableMetadata, remove.fieldNames())
+    }
+
+    val columnsToAdd = changes.collect { case add: AddColumn =>
+      (checkColumnName(add.fieldNames()), sparkSqlToJavaDriverType(add.dataType(), protocolVersion))
+    }
+
+    if (propertiesToAdd.nonEmpty) {
+      val setOptionsStatement = propertiesToAdd.foldLeft(
+        SchemaBuilder.alterTable(keyspace, table).asInstanceOf[AlterTableWithOptionsEnd]) { case (alter, prop) =>
+        alter.withOption(prop.property(), parseProperty(prop.value()))
+      }
+      connector.withSessionDo(_.execute(setOptionsStatement.asCql()))
+    }
+
+    if (columnsToRemove.nonEmpty) {
+      val dropColumnsStatement = SchemaBuilder.alterTable(keyspace, table).dropColumns(columnsToRemove: _*)
+      connector.withSessionDo(_.execute(dropColumnsStatement.asCql()))
+    }
+
+    if (columnsToAdd.nonEmpty) {
+      val addColumnStatement = columnsToAdd.foldRight(
+        SchemaBuilder.alterTable(keyspace, table).asInstanceOf[AlterTableAddColumn]
+      ) { case ((colName, dataType), alterBuilder) =>
+        alterBuilder.addColumn(colName, dataType)
+      }.asInstanceOf[AlterTableAddColumnEnd]
+      connector.withSessionDo(_.execute(addColumnStatement.asCql()))
+    }
```

Comment:
By issuing these alters as a sequence of discrete statements rather than a single monolith don't we introduce the possibility of partial application and thus leaving the underlying table in some unexpected state?

Reply:
From the JavaDoc I wrote :)
```
    * The API expects that if any change is rejected, we  should not apply
    * any changes. This is basically impossible for us since we cannot batch
    * together all of our DDL changes into a single CQL statement. To ameliorate
    * this we perform as many checks as we can up front to try to avoid beginning
    * our DDL requests which we know will fail.
```

Reply:
What, like I'm supposed to _read docs_ now?  Whatever... ;)